File: solution/2300-2399/2398.Maximum Number of Robots Within Budget/Solution.cpp
class Solution {
public:
    int maximumRobots(vector<int>& chargeTimes, vector<int>& runningCosts, long long budget) {
        deque<int> q;
        long long s = 0;
        int ans = 0;
        int n = chargeTimes.size();
        for (int l = 0, r = 0; r < n; ++r) {
            s += runningCosts[r];
            while (q.size() && chargeTimes[q.back()] <= chargeTimes[r]) {
                q.pop_back();
            }
            q.push_back(r);
            while (q.size() && (r - l + 1) * s + chargeTimes[q.front()] > budget) {
                if (q.front() == l) {
                    q.pop_front();
                }
                s -= runningCosts[l++];
            }
            ans = max(ans, r - l + 1);
        }
        return ans;
    }
};

File: solution/2300-2399/2398.Maximum Number of Robots Within Budget/Solution.go
func maximumRobots(chargeTimes []int, runningCosts []int, budget int64) int {
	s := int64(0)
	ans, j := 0, 0
	q := []int{}
	for i, a := range chargeTimes {
		for len(q) > 0 && chargeTimes[q[len(q)-1]] <= a {
			q = q[:len(q)-1]
		}
		q = append(q, i)
		s += int64(runningCosts[i])
		for len(q) > 0 && int64(chargeTimes[q[0]])+int64(i-j+1)*s > budget {
			if q[0] == j {
				q = q[1:]
			}
			s -= int64(runningCosts[j])
			j++
		}
		ans = max(ans, i-j+1)
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}